### DISCUSSION TITLE
Error using MaterialRealVectorValueAux

### DISCUSSION THREAD

**Discussion by crb3874** [idaholab/moose#19638] [Q&A General] [answered]
At a bit of a loss with this - 

I have a custom Material class `InterpolatedMGXSMaterial` which produces material property values using interpolation tables read in from a csv file. I'm getting incorrect results from simulations using it, when compared with identical simulations with directly provided material properties. The only difference between these simulations is how the material properties are being calculated, so I'm trying to debug my `InterpolatedMGXSMaterial` using `MaterialRealVectorValueAux`. 

**The issue: When using `MaterialRealVectorValueAux`, I get the error "Material has no property named: diffusivity" **

My material properties are declared in `InterpolatedMGXSMaterial.h` as

```
MaterialProperty<RealEigenVector> & _diffusivity;
```

and in `InterpolatedMGXSMaterial.C` I have

```
InterpolatedMGXSMaterial::InterpolatedMGXSMaterial(const InputParameters & parameters), 
... , 
_diffusivity(declareProperty<RealEigenVector>("diffusivity")),
 ... {...}
```

For debugging, I'm attempting to fill an AuxVariable with the element-wise values of diffusivity at the zeroth index.

Within my input file, I have
```
[AuxVariables]
...
  [diff_fast]
    family = MONOMIAL
    order = CONSTANT
  []
...
[]
[AuxKernels]
...
  [copy_diff]
    type = MaterialRealVectorValueAux
    property = diffusivity
    variable = diff_fast
    component = 0
    execute_on = final
  []
...
[]
```
When I run the simulation with these extra options, I get the error "Material has no property named: diffusivity". The strange thing is that I'm using custom Kernels as well for my simulation the directly reference the material property as "diffusivity" with no issue. 

This seems like something with a really simple fix, but as far as I can tell I've properly defined the material property, so I'm not sure what the issue is. Any thoughts?

Thanks,

Colin

**Comment by loganharbour** [comment, marked as answer]
`MaterialRealVectorValueAux` is for vector materials. You need one for array materials. I can write one up real quick.

**Reply by crb3874** [reply]
Thank you for the reply, Logan. It seems obvious now that you've said it. I may be able to write up a `MaterialRealArrayValueAux` on my own - I'll follow up with whether that fixes the issue.

**Reply by crb3874** [reply]
This seems to have fixed my issue - I wrote up a quick `MaterialRealArrayValueAux` and I seem to be able to copy my material properties to an AuxVariable. Thanks for the help!

**Reply by GiudGiud** [reply]
Hello

Would you mind making a PR to the framework to add this?
This involves adding a test and a basic documentation page if you havent done this before

Guillaume

**Reply by loganharbour** [reply]
This ended up being a little more complicated than expected - I'm almost done with it. Refs #19639

**Reply by crb3874** [reply]
One issue that came up is that attempting to check for the provided index being out of bounds for the array material property (from `this->_prop[this->_qp].size()`) in the constructor results in a segfault, but doing the same error checking in `getRealValue()` works fine. Is this because the call hierarchy in MOOSE sets up the AuxKernels before Materials, so the MaterialProperty isn't initialized until after the constructor is called for `MaterialRealArrayValueAux`?

**Reply by loganharbour** [reply]
Take a look at #19640. This adds `MaterialRealEigenVectorAux` (for a single value to a regular aux variable) and `VectorMaterialRealEigenVectorAux` (for all values to an array aux variable)

Specifically, https://github.com/idaholab/moose/blob/e5cdca1326fe035686bd96a09b4c3fc63bbd146e/framework/src/auxkernels/MaterialRealEigenVectorAux.C. You can't check upon construction because you're right, the materials aren't initialized yet. The check has to be done when getting the value.

**Reply by crb3874** [reply]
Makes sense. Thank you for adding this functionality!
